src/index.ts

```diff
@@ -148,14 +169,26 @@ async function webtransportBiDiStreamToStream (bidiStream: any, streamId: string
       }
     })(),
     sink: async function (source: Source<Uint8Array | Uint8ArrayList>) {
-      for await (const chunks of source) {
-        if (chunks.constructor === Uint8Array) {
-          await writer.write(chunks)
-        } else {
-          for (const buf of chunks) {
-            await writer.write(buf)
+      if (sinkSunk) {
+        throw errCode(new Error('sink already called on stream'), ERR_DOUBLE_SINK)
+      }
+      sinkSunk = true
+      try {
+        for await (const chunks of source) {
+          if (chunks.constructor === Uint8Array) {
+            await writer.write(chunks)
+          } else {
+            for (const buf of chunks) {
+              await writer.write(buf)
+            }
           }
         }
+      } finally {
+        // Defer stream.closeWrite a bit since there's a current bug where
+        // sometimes writes are not flushed properly. We may be able to remove
+        // this in the future, but we should run the tests many times to
+        // confirm.
```

Comment:
Is this a bug in Chrome?

Reply:
I would think so. I want to make a minimal repro (without the libp2p stack) and file it. The documentation of close says it should flush writes. So this shouldn't be needed.

Reply:
Hmm, actually I'm looking at a qlog, and I think we do end up flushing the bytes. This might be a different issue actually.

Reply:
https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/ping/ping.go#L78

We are looping between reading and echoing the bytes back. If the writer closes their side of the stream, Go will get an EOF on line 78. It will error and call stream reset on line 74. The reset is the error I'm seeing. I thought it was because we didn't finish writing our bytes before closing, but it's actually because we didn't keep doing a ping.

Reply:
I've changed this to wait for the pong before closing the write side (I basically reimplemented the ping protocol). We don't have a formal spec on the ping protocol, but here's how I think it should be defined:

For server:
1. Read 32 bytes
2. Echo those 32 bytes back to the stream.
3. Repeat
If anything fails, stream reset/close (up to the implementation), but at this point we're done with the ping protocol.

For the client:
1. Create 32 bytes of random data.
2. Write those 32 bytes to the stream.
3. Read and wait until you get those exact 32 bytes back.
4. Measure the time between 2 and after 3. This is the RTT.
5. Go to 1 or exit.

The existing JS client implementation instead does:
1. Create 32 bytes of random data.
2. Write those 32 bytes to the stream.
3. Close the write side
4. Read and wait until you get those exact 32 bytes back.
5. Measure the time between 2 and after 3. This is the RTT.
6. exit

If Go could flush the stream after the server step 2, I think the JS client impl would be fine. But afaict there's no way to flush a stream without closing it. We could close the stream instead of resetting it. That would also solve this problem. 

I think the JS client implementation should probably be changed, since it wants to measure the RTT while still in the protocol, and as soon as it closes the write side it leaves the protocol (which is why Go resets the stream here).

Reply:
Maybe Go should close the stream instead of resetting it. Definitely an easier change.

Reply:
https://github.com/libp2p/go-libp2p/pull/1853 Done on the Go side.